trieste/space.py

```diff
@@ -329,3 +346,124 @@ def __mul__(self, other: Box) -> Box:
+        """
+
+        tags = [tag for (tag, space) in spaces]
+        tf.debugging.assert_equal(
+            len(tags),
+            len(set(tags)),
+            message="subspace names must be unique",
+        )
+
+        self._tags = tags
+        self._spaces = dict(zip(self._tags, [space for (tag, space) in spaces]))
+
+        subspace_sizes = [self._spaces[tag].dimension for tag in self._tags]
+        self._subspace_sizes = dict(zip(self._tags, subspace_sizes))
+        self._subspace_starting_indicies = dict(
+            zip(self._tags, tf.cumsum(subspace_sizes, exclusive=True))
+        )
+        self._dimension = tf.reduce_sum(subspace_sizes)
+
+    def __repr__(self) -> str:
+        """"""
+        return f"""DecomposableSearchSpace(
+        {[space.__repr__()+"," for space in self._spaces]})
+        """
+
+    @property
+    def subspace_tags(self) -> list[str]:
+        """Return the names of the subspaces contained in this product space."""
+        return self._tags
+
+    @property
+    def dimension(self) -> TensorType:
+        """The input dimensionality of this product search space."""
+        return self._dimension
+
+    def get_subspace(self, tag: str) -> SearchSpace:
+        """Return the domain of a particular subspace."""
+        return self._spaces[tag]
+
+    def get_subspace_component(self, tag: str, values: TensorType) -> TensorType:
+        """
+        Returns the components of ``values`` lying in a particular subspace.
+
+        :param value: Points from the :class:`DecomposableSearchSpace` of shape [N,D].
+        :return: The context components of ``values` of shape [M, D], where
+            M is the dimensionality of the specified subspace.
+        """
+
+        starting_index_of_subspace = self._subspace_starting_indicies[tag]
+        ending_index_of_subspace = starting_index_of_subspace + self._subspace_sizes[tag]
+        return values[:, starting_index_of_subspace:ending_index_of_subspace]
+
+    def __contains__(self, value: TensorType) -> bool | TensorType:
+        """
+        Return `True` if ``value`` is a member of this search space, else `False`. A point is a
+        member if each of its subspace components lie in each subspace.
+
+        Recall that individual points in product spaces are represented with their inputs in the
+        same order as specified when initializing the space.
+
+        :param value: A point to check for membership of this :class:`SearchSpace`.
+        :return: `True` if ``value`` is a member of this search space, else `False`. May return a
+            scalar boolean `TensorType` instead of the `bool` itself.
+        :raise ValueError (or tf.errors.InvalidArgumentError): If ``value`` has a different
+            dimensionality from the search space.
+        """
+
+        tf.debugging.assert_equal(
+            tf.shape(value),
+            self.dimension,
+            message="value must have same dimensionality as search space",
```

Comment:
Adding some runtime information to logs and errors is always nice for the end users

```suggestion
            message=f"Dimensionality mismatch: space is {self.dimension}, value is {tf.shape(value}",
```

Reply:
done